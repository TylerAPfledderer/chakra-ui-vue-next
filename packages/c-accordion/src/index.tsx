export {
  CAccordion,
  CAccordionButton,
  CAccordionIcon,
  CAccordionItem,
  CAccordionPanel,
  ExpandedValues,
} from "./accordion"

export type {
  CAccordionButtonProps,
  CAccordionContext,
  CAccordionIconProps,
  CAccordionItemContext,
  CAccordionItemProps,
  CAccordionPanelProps,
  CAccordionProps,
} from "./accordion"
