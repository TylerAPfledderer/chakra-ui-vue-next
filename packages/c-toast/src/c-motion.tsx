import {
  h,
  inject,
  onMounted,
  onUpdated,
  provide,
  ref,
  defineComponent,
} from "vue"
import { createMotionState, createStyles, style } from "@motionone/dom"
import { contextId, presenceId, PresenceState } from "./c-motion.context"

const objectType = {
  type: Object,
}

export const Motion = defineComponent({
  name: "Motion",
  inheritAttrs: true,
  props: {
    tag: {
      type: String,
      default: "div",
    },
    initial: {
      type: [Object, Boolean],
    },
    animate: objectType,
    inView: objectType,
    hover: objectType,
    press: objectType,
    exit: objectType,
    inViewOptions: objectType,
    transition: objectType,
    style: objectType,
  },
  setup(props) {
    const root = ref<Element | null>(null)
    const parentState = inject(contextId, undefined)
    const presenceState = inject(presenceId, undefined) as
      | PresenceState
      | undefined

    const state = createMotionState(
      {
        ...props,
        initial:
          presenceState?.initial === false
            ? presenceState.initial
            : props.initial === true
            ? undefined
            : props.initial,
      },
      parentState
    )

    provide(contextId, state)

    onMounted(() => {
      const unmount = state.mount(root.value!)
      let height = props.animate?.height
      // if (root.value) {
      //   const el = root.value as HTMLElement
      //   const { height: initialHeight } = getComputedStyle(el)
      //   el.style.visibility = "hidden"
      //   el.style.height = props.animate?.height || "auto"
      //   const { height: targetHeight } = getComputedStyle(el)
      //   height = targetHeight
      //   el.style.height = initialHeight
      //   el.style.visibility = "visible"
      // }

      state.update({
        ...props,
        animate: {
          ...props.animate,
          // height: height,
        },
        initial: props.initial === true ? undefined : props.initial,
      })

      return unmount
    })

    let manuallyAppliedMotionStyles = false
    onUpdated(() => {
      // let height = props.exit?.height
      // if (root.value) {
      //   const el = root.value as HTMLElement
      //   const { height: initialHeight } = getComputedStyle(el)
      //   el.style.visibility = "hidden"
      //   el.style.height = props.exit?.height || "auto"
      //   const { height: targetHeight } = getComputedStyle(el)
      //   height = targetHeight
      //   el.style.height = initialHeight
      //   el.style.visibility = "visible"
      // }
      /**
       * Vue reapplies all styles every render, rather than diffing and
       * only reapplying the ones that change. This means that initially
       * calculated motion styles also get reapplied every render, leading
       * to incorrect animation origins.
       *
       * To prevent this, once an element is mounted we hand over these
       * styles to Motion. This will currently still lead to a jump if interrupting
       * transforms in browsers where the number polyfill is used.
       */
      if (!manuallyAppliedMotionStyles && root.value) {
        manuallyAppliedMotionStyles = true

        const styles = createStyles(state.getTarget())
        for (const key in styles) {
          style.set(root.value, key, styles[key])
        }
      }

      state.update({
        ...props,
        animate: {
          ...props.animate,
          // height: height,
        },
        initial: props.initial === true ? undefined : props.initial,
      })
    })

    return {
      state,
      root,
      initialStyles: createStyles(state.getTarget()),
    }
  },
  render() {
    return h(
      this.tag,
      {
        ref: "root",
        style: this.state.isMounted()
          ? this.style
          : { ...this.style, ...this.initialStyles },
      },
      this.$slots.default?.()
    )
  },
})
