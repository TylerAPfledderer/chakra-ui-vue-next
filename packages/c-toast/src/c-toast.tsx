/**
 * Hey! Welcome to @chakra-ui/vue-next CToast
 *
 * The toast component is used to give feedback to users after an action has taken place
 *
 * @see Docs     https://next.vue.chakra-ui.com/c-toast
 * @see Source   https://github.com/chakra-ui/chakra-ui-vue-next/blob/main/packages/c-toast/src/c-toast/c-toast.ts
 * @see WAI-ARIA https://www.w3.org/TR/wai-aria-practices-1.2
 */

import { defineComponent, h, Fragment, PropType } from "vue"
import {
  chakra,
  DOMElements,
} from "@chakra-ui/vue-system"

export interface CToastProps {}

export const CToast = defineComponent({
    props: {
      as: {
        type: [Object, String] as PropType<DOMElements>,
        default: "div",
      },
    },
    setup(props, { slots, attrs }) {
      return () => (
        <chakra.div as={props.as} {...attrs}>
          {slots}
        </chakra.div>
      )
    },
  })
