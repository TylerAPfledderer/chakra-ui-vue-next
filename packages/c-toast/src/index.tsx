export * from "./c-toast"
export * from "./c-presence-group"
export * from "./toast-state-machine"
export * from "./toast.utils"
