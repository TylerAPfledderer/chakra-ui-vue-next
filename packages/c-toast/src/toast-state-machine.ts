/**
 * This file is a mutated version of the @zag-js/core
 * `useMachine` and `useService` hooks.
 *
 * The goal of this mutation is to us to create the toast state
 * machine context in a global window context.
 */

import type { MachineSrc, StateMachine as S } from "@zag-js/core"
import { ComputedRef, Ref, shallowRef, watch, watchEffect } from "vue"

type MachineOptions<
  TContext extends Record<string, any>,
  TState extends S.StateSchema,
  TEvent extends S.EventObject = S.AnyEventObject
> = Omit<S.HookOptions<TContext, TState, TEvent>, "context"> & {
  context?: Ref<S.UserContext<TContext>> | ComputedRef<S.UserContext<TContext>>
}

export function useService<
  TContext extends Record<string, any>,
  TState extends S.StateSchema,
  TEvent extends S.EventObject = S.AnyEventObject
>(
  machine: MachineSrc<TContext, TState, TEvent>,
  options?: MachineOptions<TContext, TState, TEvent>
) {
  const { actions, state: hydratedState, context } = options ?? {}

  const _machine = typeof machine === "function" ? machine() : machine
  const service = context ? _machine.withContext(context.value) : _machine

  window.addEventListener("load", () => {
    service.start(hydratedState)

    if (service.state.can("SETUP")) {
      service.send("SETUP")
    }

    window.addEventListener("unload", () => {
      service.stop()
    })
  })

  watchEffect(() => {
    service.setOptions({ actions })
  })

  if (context) {
    watch(context, service.setContext, { deep: true })
  }

  return service
}

export function useMachine<
  TContext extends Record<string, any>,
  TState extends S.StateSchema,
  TEvent extends S.EventObject = S.AnyEventObject
>(
  machine: MachineSrc<TContext, TState, TEvent>,
  options?: Omit<S.HookOptions<TContext, TState, TEvent>, "context"> & {
    context?:
      | Ref<S.UserContext<TContext>>
      | ComputedRef<S.UserContext<TContext>>
  }
) {
  const service = useService(machine, options)
  const state = shallowRef(service.state)

  window.addEventListener("load", () => {
    const unsubscribe = service.subscribe((nextState) => {
      state.value = nextState
    })
    window.addEventListener("unload", () => {
      unsubscribe?.()
    })
  })

  return [state, service.send, service] as const
}
