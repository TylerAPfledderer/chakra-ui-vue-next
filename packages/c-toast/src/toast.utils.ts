import * as __toast__ from "@zag-js/toast"

export const placements = [
  "top-start",
  "top",
  "top-end",
  "bottom-start",
  "bottom",
  "bottom-end",
] as const

export type ToastPlacement = (typeof placements)[number]

export function getToastsByPlacement(toasts: __toast__.Service[]) {
  const result: Partial<Record<ToastPlacement, __toast__.Service[]>> = {
    "top-start": [],
    top: [],
    "top-end": [],
    "bottom-start": [],
    bottom: [],
    "bottom-end": [],
  }

  for (const toast of toasts) {
    const placement = toast.state.context.placement!
    result[placement] ||= []
    result[placement]!.push(toast)
  }

  return result
}

export { __toast__ }
