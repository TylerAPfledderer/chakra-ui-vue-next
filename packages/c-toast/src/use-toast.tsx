import * as toast from "@zag-js/toast"
import { normalizeProps, useActor, useMachine } from "@zag-js/vue"
