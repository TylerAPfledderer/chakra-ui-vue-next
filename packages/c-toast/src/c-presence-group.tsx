import {
  defineComponent,
  h,
  onBeforeUpdate,
  provide,
  Transition,
  TransitionGroup,
} from "vue"
import { mountedStates } from "@motionone/dom"
import { contextId, presenceId } from "./c-motion.context"

export interface PresenceState {
  initial?: boolean | undefined
}

let _id = 1
const createPresenceGroupContextId = () => Symbol(`presence-group-${_id}`)

const doneCallbacks = new WeakMap<Element, VoidFunction>()

function removeDoneCallback(element: Element) {
  const prevDoneCallback = doneCallbacks.get(element)
  prevDoneCallback &&
    element.removeEventListener("motioncomplete", prevDoneCallback)
  doneCallbacks.delete(element)
}

export const CPresenceGroup = defineComponent({
  name: "CPresenceGroup",
  props: {
    name: { type: String },
    initial: {
      type: Boolean,
      default: true,
    },
    exitBeforeEnter: {
      type: Boolean,
      default: false,
    },
  },
  setup(props, { slots, attrs }) {
    const state: PresenceState = { initial: props.initial }
    const presenseGroupContextId = createPresenceGroupContextId()

    provide(presenceId, state)

    function enter(element: Element) {
      const state = mountedStates.get(element)
      console.log("enter state", element, state)

      if (!state) return

      removeDoneCallback(element)
      state.setActive("exit", false)
    }

    function exit(element: Element, done: VoidFunction) {
      const state = mountedStates.get(element)
      console.log("exit state", element, state)

      if (!state) return done()

      state.setActive("exit", true)

      removeDoneCallback(element)
      doneCallbacks.set(element, done)
      element.addEventListener("motioncomplete", done)
    }

    onBeforeUpdate(() => {
      state.initial = undefined
    })

    return () => (
      <TransitionGroup onEnter={enter} onLeave={exit} css={false}>
        {slots.default?.()}
      </TransitionGroup>
    )
  },
})
