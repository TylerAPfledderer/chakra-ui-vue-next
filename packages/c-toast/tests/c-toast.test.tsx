import { h, Fragment, defineComponent } from "vue"
import { render } from "../../test-utils/src"
import { CToast } from "../src"
it("should render properly", () => {
  const { asFragment } = render(
    defineComponent(() => {
      return () => <h1>Hello World</h1>
    })
  )
  expect(asFragment()).toMatchSnapshot()
})
