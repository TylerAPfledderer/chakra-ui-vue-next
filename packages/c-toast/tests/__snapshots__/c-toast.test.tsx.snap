// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render properly 1`] = `
<DocumentFragment>
  <h1>
    Hello World
  </h1>
</DocumentFragment>
`;
